File: c/binary-search/src/binary_search.c
#include "binary_search.h"
#include <stdio.h>

int *binary_search(int value, int arr[], size_t length)
{
  if (!arr || length == 0) return NULL;
  int min = 0;
  int max = length - 1;
  while (min != max)
  {
    int mid = (min + max) / 2;
    if (value == arr[mid]) return &arr[mid];
    if (value > arr[mid])
    {
      min = mid + 1;
    }
    else
    {
      max = mid - 1;
    }
  }
  return arr[min] == value? &arr[min] : NULL;
}

File: javascript/binary-search/binary-search.js
const isSorted = arr => arr.slice().sort((a, b) => a - b).join('') === arr.join('');

class BinarySearch {
  constructor(arr) {
    if (isSorted(arr)) {
      this.array = arr;
    }
  }

  indexOf(target) {
    let min = 0;
    let max = this.array.length - 1;

    while (min <= max) {
      const mid = Math.floor((min + max) / 2);
      if (this.array[mid] === target) {
        return mid;
      }
      if (target > this.array[mid]) {
        min = mid + 1;
      }
      max = mid - 1;
    }
    return -1;
  }
}

export { BinarySearch };
